crates/stages/src/stages/headers.rs

```diff
@@ -218,6 +221,61 @@ where
 
         info!(target: "sync::stages::headers", len = downloaded_headers.len(), "Received headers");
 
+        let tip_block_number = match tip {
+            // If tip is hash and it equals to the first downloaded header's hash, we can use
+            // the block number of this header as tip.
+            BlockHashOrNumber::Hash(hash) => downloaded_headers.first().and_then(|header| {
+                if header.hash == hash {
+                    Some(header.number)
+                } else {
+                    None
+                }
+            }),
+            // If tip is number, we can just grab it and not resolve using downloaded headers.
+            BlockHashOrNumber::Number(number) => Some(number),
+        };
+
+        // Since we're syncing headers in batches, gap tip will move in reverse direction towards
+        // our local head with every iteration. To get the actual target block number we're
+        // syncing towards, we need to take into account already synced headers from the database.
+        // It is `None`, if tip didn't change and we're still downloading headers for previously
+        // calculated gap.
+        let target_block_number = if let Some(tip_block_number) = tip_block_number {
+            let local_max_block_number = tx
+                .cursor_read::<tables::CanonicalHeaders>()?
+                .last()?
+                .map(|(canonical_block, _)| canonical_block);
```

Comment:
I might be misunderstanding why we need this, but don't we have the target block in the `gap` field?

Reply:
Imagine such situation:
1. We start syncing from scratch up to a block height 1000
2. Headers stage determines the gap as `local_head = 0, target = 1000`
3. On the first iteration, stage downloads 100 headers, and we restart the node
4. Now, headers stage determines the gap as `local_head = 0, target = 900`, because we downloaded 100 headers from the end

In this scenario, the `gap.target.tip()` will point at the block 900, but in reality we're syncing towards block 1000 and already has its header in the database. With these lines, we'll grab the maximum block header from the database, and it'll be header 1000 which we downloaded before.

Reply:
ah, right!
ty